src/main/java/org/openrewrite/github/SetupJavaUpgradeJavaVersion.java

```diff
@@ -0,0 +1,73 @@
+package org.openrewrite.github;
+
+import lombok.EqualsAndHashCode;
+import lombok.Value;
+import org.openrewrite.*;
+import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.yaml.JsonPathMatcher;
+import org.openrewrite.yaml.YamlVisitor;
+import org.openrewrite.yaml.tree.Yaml;
+
+@Value
+@EqualsAndHashCode(callSuper = true)
+public class SetupJavaUpgradeJavaVersion extends Recipe {
+
+    @Option(displayName = "Minimum major Java version (defaults to 17)",
+            example = "17",
+            required = false)
+    @Nullable
+    Integer minimumJavaMajorVersion;
+
+    @Override
+    public String getDisplayName() {
+        return "Upgrade `actions/setup-java` `java-version`";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Update the Java version used by `actions/setup-java` if it is below the expected version number.";
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getSingleSourceApplicableTest() {
+        return new HasSourcePath<>(".github/workflows/*.yml");
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new UpgradeJavaVersionVisitor(
+                minimumJavaMajorVersion == null ? 17 : minimumJavaMajorVersion
+        );
+    }
+
+    private static class UpgradeJavaVersionVisitor extends YamlVisitor<ExecutionContext> {
+        private static final JsonPathMatcher javaVersion = new JsonPathMatcher("..steps[?(@.uses =~ 'actions/setup-java@v*.*')].with.java-version");
+
+        private final int minimumJavaMajorVersion;
+
+        public UpgradeJavaVersionVisitor(int minimumJavaMajorVersion) {
+            this.minimumJavaMajorVersion = minimumJavaMajorVersion;
+        }
+
+        @Override
+        public Yaml visitMappingEntry(Yaml.Mapping.Entry entry, ExecutionContext ctx) {
+            if (!javaVersion.matches(getCursor())) {
+                return super.visitMappingEntry(entry, ctx);
+            }
+
+            Yaml.Scalar currentValue = (Yaml.Scalar) entry.getValue();
+
+            // specific versions are allowed by `actions/setup-java`
+            String[] currentVersionParts = currentValue.getValue().split("\\.");
+            int currentMajorVersion = Integer.parseInt(currentVersionParts[0]);
+            if (currentMajorVersion >= minimumJavaMajorVersion) {
+                return super.visitMappingEntry(entry, ctx);
+            }
+
+            return super.visitMappingEntry(
+                    entry.withValue(currentValue.withValue(String.valueOf(minimumJavaMajorVersion))),
+                    ctx
+            );
+        }
```

Comment:
By convention we typically do the `super.visit<Element>` bit first unless there is a reason not to. 

It's also possible that the value here isn't a literal like "11", but is instead a reference to some property defined elsewhere. Something like `${{ something.JAVA_VERSION }}`. So we might get some `NumberFormatException` coming out of that `Integer.parseInt()`. I recommend catching those and returning an unmodified `e` in that case. 

```suggestion
        @Override
        public Yaml visitMappingEntry(Yaml.Mapping.Entry entry, ExecutionContext ctx) {
            Yaml.Mapping.Entry e = super.visitMappingEntry(entry, ctx);
            if (!javaVersion.matches(getCursor())) {
                return e;
            }

            Yaml.Scalar currentValue = (Yaml.Scalar) e.getValue();

            // specific versions are allowed by `actions/setup-java`
            String[] currentVersionParts = currentValue.getValue().split("\\.");
            int currentMajorVersion = Integer.parseInt(currentVersionParts[0]);
            if (currentMajorVersion <= minimumJavaMajorVersion) {
                e = e.withValue(currentValue.withValue(String.valueOf(minimumJavaMajorVersion)));
            }

            return e;
        }
```

Reply:
> By convention we typically do the super.visit<Element> bit first unless there is a reason not to.

I'd based this on `ActionsSetupJavaAdoptOpenJDKToTemurinTest`; not sure if there was a reason it was done that way there or if it's just inconsistent.

I'll push an update to handle the exception case!

Reply:
Ah, because `super.visitMappingEntry` returns `Yaml` I think the code I'd based this on did it this way to avoid the cast / assumption (i.e. `Yaml.Mapping.Entry e = (Yaml.Mapping.Entry) super.visitMappingEntry(entry, ctx);`).

Reply:
The cast is safe unless this visitor would change it. Since every code path within your `visitMappingEntry` returns a `Yaml.Mapping.Entry` the cast is safe